lib/assert.js

```diff
@@ -476,23 +474,37 @@ function partiallyCompareArrayBuffersOrViews(actual, expected) {
 
 function partiallyCompareSets(actual, expected, comparedObjects) {
   if (SetPrototypeGetSize(expected) > SetPrototypeGetSize(actual)) {
-    return false; // `expected` can't be a subset if it has more elements
+    return false;
   }
 
   if (isDeepEqual === undefined) lazyLoadComparison();
 
-  const actualArray = ArrayFrom(FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual));
   const expectedIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], expected);
-  const usedIndices = new SafeSet();
+  let actualSet;
+
+  for (const expectedItem of expectedIterator) {
+    let foundMatch = false;
 
-  expectedIteration: for (const expectedItem of expectedIterator) {
-    for (let actualIdx = 0; actualIdx < actualArray.length; actualIdx++) {
-      if (!usedIndices.has(actualIdx) && isDeepStrictEqual(actualArray[actualIdx], expectedItem)) {
-        usedIndices.add(actualIdx);
-        continue expectedIteration;
+    // Check for primitives first to avoid useless loops
+    if (isPrimitive(expectedItem)) {
+      if (actual.has(expectedItem)) {
+        foundMatch = true;
+      }
+    } else {
+      actualSet ??= new SafeSet(FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual));
```

Comment:
This could be optimised for mixed sets (objects and primitives) by only adding objects. That is how it's implemented in the deepStrictEqual implementation and it could be used as blueprint :)